sensor/common/compliance/node_inventory_handler_impl.go

```diff
@@ -73,41 +80,84 @@ func (c *nodeInventoryHandlerImpl) ProcessMessage(_ *central.MsgToSensor) error
+			if !ok {
+				c.stopper.Flow().StopWithError(errInputChanClosed)
 				return
-			case inventory, ok := <-c.inventories:
-				if !ok {
-					c.stopper.Flow().StopWithError(errInputChanClosed)
-					return
-				}
-				if !c.centralReady.IsDone() {
-					// TODO(ROX-13164): Reply with NACK to compliance
-					log.Warnf("Received node inventory but Central is unavailable")
-					continue
-				}
-				if inventory == nil {
-					log.Warnf("Received nil node inventory: not sending to Central")
-					break
-				}
-				if nodeID, err := c.nodeMatcher.GetNodeID(inventory.GetNodeName()); err != nil {
-					log.Warnf("Node '%s' unknown to sensor: not sending node inventory to Central", inventory.GetNodeName())
-				} else {
-					inventory.NodeId = nodeID
-					metrics.ObserveReceivedNodeInventory(inventory)
-					log.Debugf("Mapping node inventory name '%s' to Node ID '%s'", inventory.GetNodeName(), nodeID)
-					c.sendNodeInventory(toC, inventory)
-				}
+			}
+			if !c.centralReady.IsDone() {
+				log.Warnf("Received NodeInventory but Central is not reachable. Requesting Compliance to resend NodeInventory later")
+				c.sendNackToCompliance(toCompliance, inventory)
+				continue
+			}
+			if inventory == nil {
+				log.Warnf("Received nil node inventory: not sending to Central")
+				break
+			}
+			if nodeID, err := c.nodeMatcher.GetNodeID(inventory.GetNodeName()); err != nil {
+				log.Infof("Sending NACK to compliance after receiving unknown NodeInventory with ID %s", inventory.GetNodeId())
+				c.sendNackToCompliance(toCompliance, inventory)
+			} else {
+				inventory.NodeId = nodeID
+				metrics.ObserveReceivedNodeInventory(inventory)
+				log.Debugf("Mapping NodeInventory name '%s' to Node ID '%s'", inventory.GetNodeName(), nodeID)
+				c.sendNodeInventory(toCentral, inventory)
+				log.Debugf("Sending ACK to compliance for NodeInventory with ID %s", inventory.GetNodeId())
+				c.sendAckToCompliance(toCompliance, inventory)
 			}
 		}
-	}()
-	return toC
+	}
+}
+
+func (c *nodeInventoryHandlerImpl) sendNackToCompliance(complianceC chan<- *common.MessageToComplianceWithAddress, inventory *storage.NodeInventory) {
+	if inventory == nil {
+		return
+	}
+	complianceC <- &common.MessageToComplianceWithAddress{
+		Msg: &sensor.MsgToCompliance{
+			Msg: &sensor.MsgToCompliance_Nack{
+				Nack: &sensor.MsgToCompliance_NodeInventoryNack{
+					NodeId: inventory.GetNodeId(),
+				},
+			},
+		},
+		Hostname:  inventory.GetNodeName(),
+		Broadcast: false,
+	}
+}
+
+func (c *nodeInventoryHandlerImpl) sendAckToCompliance(complianceC chan<- *common.MessageToComplianceWithAddress, inventory *storage.NodeInventory) {
+	if inventory == nil {
+		return
+	}
+	complianceC <- &common.MessageToComplianceWithAddress{
+		Msg: &sensor.MsgToCompliance{
+			Msg: &sensor.MsgToCompliance_Ack{
+				Ack: &sensor.MsgToCompliance_NodeInventoryAck{
+					NodeId: inventory.GetNodeId(),
+				},
+			},
+		},
+		Hostname:  inventory.GetNodeName(),
+		Broadcast: false,
+	}
```

Comment:
This func may be removed as well when we remove the call to it in line 123.

Reply:
Done